test/e2e/apps/sidecarset.go

```diff
@@ -1576,5 +1577,115 @@ var _ = SIGDescribe("SidecarSet", func() {
+
+			// check sidecar original revisions
+			podStable, err := tester.GetSelectorPods(ns, &metav1.LabelSelector{MatchLabels: deploymentStable.Spec.Template.Labels})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			gomega.Expect(podStable).To(gomega.HaveLen(1))
+			stableNum, canaryNum = calculateSidecarImages(podStable)
+			gomega.Expect(stableNum).To(gomega.Equal(1))
+			gomega.Expect(canaryNum).To(gomega.Equal(0))
+			podCanary, err := tester.GetSelectorPods(ns, &metav1.LabelSelector{MatchLabels: deploymentCanary.Spec.Template.Labels})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			gomega.Expect(podCanary).To(gomega.HaveLen(1))
+			stableNum, canaryNum = calculateSidecarImages(podCanary)
+			gomega.Expect(stableNum).To(gomega.Equal(1))
+			gomega.Expect(canaryNum).To(gomega.Equal(0))
+			ginkgo.By(fmt.Sprintf("All pods are injected with stable sidecar"))
+
+			// canary update sidecarSet
+			sidecarSetIn.Spec.Containers[0].Image = nginxName(canaryTag)
+			sidecarSetIn.Spec.Containers[0].Env = []corev1.EnvVar{{Name: "SOME_ENV", Value: "SOME_VAL"}} // make in-place upgrade impossible
+			sidecarSetIn.Spec.UpdateStrategy.Selector = &metav1.LabelSelector{
+				MatchLabels: map[string]string{"version": "canary"},
+			}
+			sidecarSetIn.Spec.InjectionStrategy.Revision = &appsv1alpha1.SidecarSetInjectRevision{CustomVersion: ptr.To("0")}
+			sidecarSetIn.Labels = map[string]string{
+				appsv1alpha1.SidecarSetCustomVersionLabel: "1",
+			}
+			tester.UpdateSidecarSet(sidecarSetIn)
+			except := &appsv1alpha1.SidecarSetStatus{
+				MatchedPods:      2,
+				UpdatedPods:      0,
+				UpdatedReadyPods: 0,
+				ReadyPods:        2,
+			}
+			tester.WaitForSidecarSetUpgradeComplete(sidecarSetIn, except)
+			ginkgo.By(fmt.Sprintf("Sidecarset Updated"))
+
+			// scale up deployments
+			deploymentStable.Spec.Replicas = ptr.To(int32(2))
+			deploymentCanary.Spec.Replicas = ptr.To(int32(2))
+			tester.UpdateDeployment(deploymentStable)
+			tester.UpdateDeployment(deploymentCanary)
+			tester.WaitForDeploymentRunning(deploymentStable)
+			tester.WaitForDeploymentRunning(deploymentCanary)
+
+			// check sidecar versions
+			podStable, err = tester.GetSelectorPods(ns, &metav1.LabelSelector{MatchLabels: deploymentStable.Spec.Template.Labels})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			gomega.Expect(podStable).To(gomega.HaveLen(2))
+			stableNum, canaryNum = calculateSidecarImages(podStable)
+			gomega.Expect(stableNum).To(gomega.Equal(2))
+			gomega.Expect(canaryNum).To(gomega.Equal(0))
+			podCanary, err = tester.GetSelectorPods(ns, &metav1.LabelSelector{MatchLabels: deploymentCanary.Spec.Template.Labels})
+			gomega.Expect(err).NotTo(gomega.HaveOccurred())
+			gomega.Expect(podCanary).To(gomega.HaveLen(2))
+			stableNum, canaryNum = calculateSidecarImages(podCanary)
+			gomega.Expect(stableNum).To(gomega.Equal(1))
+			gomega.Expect(canaryNum).To(gomega.Equal(1))
+			ginkgo.By(fmt.Sprintf("All pods are injected with stable sidecar"))
```

Comment:
this log is incorrect since newly created canary pods are injected with canary sidecar